module/zfs/dmu_objset.c

```diff
@@ -1311,6 +1313,7 @@ dmu_objset_upgrade_task_cb(void *data)
 	os->os_upgrade_exit = B_TRUE;
 	os->os_upgrade_id = 0;
 	mutex_exit(&os->os_upgrade_lock);
+	dsl_dataset_long_rele(dmu_objset_ds(os), upgrade_tag);
```

Comment:
There's one case in `dmu_objset_upgrade_stop()` when you need to drop the long hold.  When `taskq_cancel_id()` return's 0 it indicates the taskqid was able to be be successfully canceled before it was executed by a thread.  In which case you'll need to drop the long hold there.  In both the ENOENT and EBUSY error cases the callback will either have run or be currently running so there's no problem there.

Reply:
Thank you, I'll fix this.